gobblin-data-management/src/main/java/gobblin/data/management/copy/hive/HiveCopyEntityHelper.java

```diff
@@ -0,0 +1,554 @@
+        priority = addDeregisterSteps(copyableFiles, deregisterFileSet, priority, this.targetTable,
+            partitionEntry.getValue());
+      }
+
+    } else {
+      copyableFiles.addAll(getCopyEntitiesForUnpartitionedTable());
+    }
+    return copyableFiles;
+  }
+
+  /**
+   * Takes a path with tokens {@link #databaseToken} or {@link #tableToken} and replaces these tokens with the actual
+   * database names and table name. For example, if db is myDatabase, table is myTable, then /data/$DB/$TABLE will be
+   * resolved to /data/myDatabase/myTable.
+   */
+  private Path resolvePath(String pattern) {
+    pattern = pattern.replaceAll(databaseToken, this.dataset.table.getDbName());
+    if (pattern.contains(tableToken)) {
+      pattern = pattern.replaceAll(tableToken, this.dataset.table.getTableName());
+      return new Path(pattern);
+    } else {
+      return new Path(pattern, this.dataset.table.getTableName());
+    }
+  }
+
+  private Table getTargetTable(Table originTable, Path targetLocation) throws IOException {
+    try {
+      Table targetTable = originTable.copy();
+      targetTable.setDbName(this.targetDatabase);
+      targetTable.setDataLocation(targetLocation);
+      return targetTable;
+    } catch (HiveException he) {
+      throw new IOException(he);
+    }
+  }
+
+  private Partition getTargetPartition(Partition originPartition, Path targetLocation) throws IOException {
+    try {
+      Partition targetPartition = new Partition(this.targetTable, originPartition.getTPartition().deepCopy());
+      targetPartition.getTable().setDbName(this.targetDatabase);
+      targetPartition.setLocation(targetLocation.toString());
+      return targetPartition;
+    } catch (HiveException he) {
+      throw new IOException(he);
+    }
+  }
+
+  /**
+   * Creates {@link CopyEntity}s for a partition.
+   */
+  @AllArgsConstructor
+  private class PartitionCopy {
+
+    private final Partition partition;
+
+    private List<CopyEntity> getCopyEntities()
+        throws IOException {
+
+      int stepPriority = 0;
+      String fileSet = gson.toJson(this.partition.getValues());
+
+      List<CopyEntity> copyEntities = Lists.newArrayList();
+
+      stepPriority = addSharedSteps(copyEntities, fileSet, stepPriority);
+
+      InputFormat<?, ?> inputFormat = HiveDataset.getInputFormat(this.partition.getTPartition().getSd());
+
+      Optional<Partition> targetPartition = Optional.fromNullable(targetPartitions.get(partition.getValues()));
+
+      if (targetPartition.isPresent()) {
+        targetPartitions.remove(partition.getValues());
+        try {
+          checkPartitionCompatibility(partition, targetPartition.get());
+        } catch (IOException ioe) {
+          if (existingEntityPolicy == ExistingEntityPolicy.ABORT) {
+            log.error("Source and target partitions are not compatible. Aborting copy of partition " + this.partition);
+            return Lists.newArrayList();
+          }
```

Comment:
Do you need to check if `ExistingEntityPolicy` equals to `REPLACE`? It is dangerous to make `REPLACE` default here.